Support for TCP Searches

sveseli opened the pull request:
This PR has been modified from the original (support for direct tcp connections) to include support for tcp searches.

Server side changes:
* Servers have ability to respond to tcp searches on the server port 

Client side changes:
* By default, only udp channel searches are performed as before
* Specifying EPICS_PVA_NAME_SERVERS variable will result in tcp search queries to the listed HOST:PORT addresses, in addition to the regular udp search

Here are some examples of how things work. In terminal 1, we run test server using non-default ports:

<pre>
daq-dss02> EPICS_PVA_SERVER_PORT=11111 EPICS_PVA_BROADCAST_PORT=22222 ./modules/pvAccess/testApp/O.linux-x86_64/testServer
pvAccess Server v7.1.8-SNAPSHOT
Active configuration (w/ defaults)
EPICS_PVAS_INTF_ADDR_LIST = 0.0.0.0:11111
EPICS_PVAS_BEACON_ADDR_LIST = 
EPICS_PVAS_AUTO_BEACON_ADDR_LIST = YES
EPICS_PVAS_BEACON_PERIOD = 15
EPICS_PVAS_BROADCAST_PORT = 22222
EPICS_PVAS_SERVER_PORT = 11111
EPICS_PVAS_PROVIDER_NAMES = local
...
</pre>

In terminal 2, on a second machine, we run client. If we do not specify anything, client fails to connect:

<pre>
daq-qss21> pvget testCounter
Timeout
testCounter
</pre>

If we specify broadcast port, channel connects using udp discovery:

<pre>
daq-qss21> EPICS_PVA_BROADCAST_PORT=22222 pvget testCounter
testCounter 2024-03-22 14:01:33.332  27 
daq-qss21> EPICS_PVA_BROADCAST_PORT=22222 pvget testCounter
testCounter 2024-03-22 14:01:33.332  27 
daq-qss21> EPICS_PVA_BROADCAST_PORT=22222 pvget -d testCounter
2024-03-22T14:01:41.838 Configured PVA address list: 
2024-03-22T14:01:41.838 Configured server port: 5075
2024-03-22T14:01:41.838 Configured name server address list: 
2024-03-22T14:01:41.838 Configured broadcast port: 22222
...
2024-03-22T14:01:41.839 Server address decoded as 10.6.13.8:11111, transport type is udp
...
2024-03-22T14:01:41.841 Connected to PVA server: 10.6.13.8:11111.
2024-03-22T14:01:41.841 Unregistering search instance: testCounter
testCounter 2024-03-22 14:01:41.333  35 
...
</pre>

If we specify EPICS_PVA_NAME_SERVERS variable, tcp search will result in channel connection:

<pre>
daq-qss21> EPICS_PVA_NAME_SERVERS=daq-dss02:11111 pvget testCounter
testCounter 2024-03-22 14:14:16.483  790 
daq-qss21> EPICS_PVA_NAME_SERVERS=daq-dss02:11111 pvget -d testCounter
2024-03-22T14:14:23.016 Configured PVA address list: 
2024-03-22T14:14:23.016 Configured server port: 5075
2024-03-22T14:14:23.016 Configured name server address list: daq-dss02:11111
2024-03-22T14:14:23.016 Configured broadcast port: 5076
2024-03-22T14:14:23.018 Creating datagram socket from: 0.0.0.0:53776.
...
2024-03-22T14:14:23.019 Getting name server transport for address 10.6.13.8:11111
...
2024-03-22T14:14:23.021 Searching for channel: testCounter
2024-03-22T14:14:23.021 Server address decoded as 10.6.13.8:11111, transport type is tcp
...
2024-03-22T14:14:23.021 Connecting to PVA server: 10.6.13.8:11111.
...
testCounter 2024-03-22 14:14:22.485  796 
...
</pre>

sveseli committed:
add ability to connect to server channels when UDP search is not feasible (e.g., if the connection has to go through firewalls); this can be done in those cases where server address is simply known, using combination of EPICS_PVA_ADDR_LIST and EPICS_PVA_SERVER_PORT variables

sveseli committed:
remove un-needed code

sveseli committed:
allow control of the UDP sender port via the EPICS_PVA_UDP_SENDER_PORT environment variable on both server and client side

kasemir reviewed (approved):
I can't comment on the PVA C++ specifics, I'm simply not familiar with it.
But as for configuring direct TCP connections to avoid the UDP searches:
PVXS and the new Java implementation use  EPICS_PVA_NAME_SERVERS to set a list of TCP addresses. The "name server" in here is a bit misleading. You could of course list an actual name server IP, but to the client it just means that it sends a search request via TCP. The message handling code on the server is really the same, no matter if a message is received via UDP or TCP.
In CA, I think it was similar: EPICS_CA_NAME_SERVERS lists TCP addresses for searches.

So with EPICS_PVA_NAME_SERVERS="IP1 IP2 IP3" the client will send searches to those 3 IP addresses. A name server might reply with a search response that lists the actual IOC, but if the address is already an IOC, that IOC can return a search response with zero for the IP address to indicate: Continue on this TCP connection for the data exchange.
In more detail, EPICS_PVA_NAME_SERVERS may contain "IP:port" to list a TCP address with port. See https://github.com/epics-base/pvAccessCPP/wiki/Protocol-Messages#cmd_search-0x03 for details.

kasemir commented:
So with PVXS and the new java lib, this is how you would configure TCP-only searches:
```
EPICS_PVA_NAME_SERVERS="IP1:PORT1 IP2:PORT2"
EPICS_PVA_AUTO_ADDR_LIST=NO
EPICS_PVA_ADDR_LIST=""
```

If only setting EPICS_PVA_NAME_SERVERS, it will contact those TCP addresses plus still send UDP searches.

anjohnson commented:
@kasemir is there an existing PVA name server?

@sveseli is looking to write one based on pvaPy if there isn't one already out there, but we'd prefer to use something that already exists.

mdavidsaver commented:
I feel bound to point out that PVXS already has a more complete "direct connection" (aka. bypassing the search phase entirely).  One usage is in [`pvxlist`](https://github.com/mdavidsaver/pvxs/blob/93ab81c1532b7078aa261d8608db713898b545b6/tools/list.cpp#L160) to connect to the un-searchable `server` PV.  Unlike `pvlist` in pvAccessCPP, `pvxlist` is implemented using public API.

Also, fair warning.  As I recall pvAccessCPP does not correctly handle reconnection of "direct" channels.  So imo. it is not suitable for use with subscriptions specifically and may cause problems with long running clients generally.

sveseli commented:
From what I have seen, before this PR direct connection to channels was not possible at all, as the code relied entirely on UDP searches, and if those are not resulting in channel discovery, there would be no connection made.  In my tests, this PR handles monitor re-connections correctly, as long as the server comes up on the same machine and same port, regardless of whether UDP search is available or not.

mdavidsaver reviewed (changes requested):
Please add some test converge.

kasemir commented:
> is there an existing PVA name server?

The Java PVA lib contains a command line demo, https://github.com/ControlSystemStudio/phoebus/blob/master/core/pva/src/test/java/org/epics/pva/server/SearchMonitorDemo.java

It takes list of PVs and their IP:port on the command line and then replies to searches for those PVs with the provided name. That's of course only usable as a demo for one or two PVs, but it allows testing `EPICS_PVA_NAME_SERVERS`, so clients go to the name server via TCP, get the IP:port of the IOC and then connect via TCP to the IOC.

The plan was somewhat like this:

Extend the channel finder's IOC tool (reccaster?) to provide the TCP port. I think that's been completed.
Channel finder will thus not only know the PV names but also the TCP IP and port.
Next build a name server that starts like that demo but instead of taking info for 1 or 2 PVs from the command line it queries the channel finder.

anjohnson commented:
Thanks Kay. Given that the PVA protocol has the ability for a name-server to query the servers for their complete list of PV names I don't think an oracle for PV names would be essential, although it might be good for performance reasons.

There are 2 modes that I could see it using, maybe both at once: When the name-server sees a search request for a PV name that it doesn't recognize it would request that name through its client-side API, and would handle any response by asking the server that answered for all of its PVs. Alternatively/also it could monitor for beacons from new servers coming up, and proactively ask them for their names. The first mode would be needed for it to work with dynamic servers which could add new PVs at runtime, if it sees a response from a known server it would refresh the list of names for that server.

I'm not quite sure how to handle dynamic servers that can drop PV names, somehow the name-server needs to know that's happened. I suspect the pvlist request can't be issued as a monitor, although that would be really helpful.

It's important for the server to cache the PV names requested that it _doesn't_ have a server for so it doesn't replicate every random request (the CA name-server has one). Supporting a configurable list of regexp's that match names to be forwarded and/or to not be forwarded might be a reasonable alternative to the Channel Finder to help with that.

anjohnson commented:
@mdavidsaver Unfortunately APS would need several FTE's of effort to convert all our DAQ software to the PVXS API, and we also rely on the plugin abilities currently unique to the pvAccess and pvDatabase libraries to be able to handle the high data rates coming from those DAQ systems. The data distributor plugin lets us accept updates from one data stream and fan them out to multiple clients to be processed in parallel. This PR provides the ability to connect through firewalls.

I do agree with the need for tests.

kasemir commented:
Andrew, you're correct, a name server might either rely on some type of name database (Channel Finder, Oracle, ...), or it might build that database itself. It could issue "pvlist" requests, or operate similar to the gateway by sending its own search request and then memorizing the reply.

@shroffk those are options we could consider if we ever get back to working on a name server.

shroffk commented:
We could consider a combination of those actions

I have a very basic name server which uses ChannelFinder populated with recsync and the PVA port.
https://github.com/ChannelFinder/cfNameserver

I had imagined that the fall back mechanism for the name server if it fails to find the name resolution in CF would be to do its own name resolution search.

sveseli committed:
add tests for udp search, name server search, and direct connection

sveseli committed:
set connected status on state change

sveseli committed:
add method for retrieving search server port

sveseli committed:
add new tests to Makefile

sveseli committed:
fix typo

sveseli committed:
add name server search

sveseli committed:
add protocol constants

sveseli committed:
use constant for type

sveseli committed:
support for name server search

sveseli committed:
use constant for tcp transport

sveseli committed:
add interface for getting name server transport

sveseli committed:
add support for tcp search

sveseli committed:
add support for name server channel discovery

sveseli committed:
add tcp acceptor for channel search

sveseli committed:
support for tcp channel search

sveseli committed:
introduce new interfaces for managing tcp transport

sveseli committed:
use different channel priority for name server searches; move name server transport to channel context

sveseli committed:
stop search if channel is found via direct connection

sveseli committed:
use default address list if provided string is empty

sveseli committed:
allow initializing server context with different set of response handlers

sveseli committed:
expose additional header

sveseli committed:
fix logging typo

sveseli committed:
add utility function to convert string to inet address

sveseli committed:
move utility functions to pvutils.cpp

sveseli committed:
add PVA name server utility

sveseli committed:
add name server utility description

sveseli committed:
add utilities for manipulating strings

sveseli committed:
add additional utility methods

sveseli committed:
add support for a static channel map read from a file

sveseli committed:
update description

sveseli committed:
send search response immediately

sveseli committed:
send name server search response immediately

sveseli committed:
merge changes for udp sender port

sveseli committed:
update for older gcc compiler

sveseli committed:
fix windows build

sveseli committed:
move base nameserver code to the library

sveseli committed:
fix build on older gcc version

sveseli committed:
fixes for mingw builds

sveseli committed:
fix few log formats

sveseli committed:
fix few codacy complaints; avoid duplicate symbold on rtems

sveseli committed:
move string constants to pvaConstants

sveseli committed:
include missing header

sveseli committed:
fix few log statement formats

sveseli committed:
add diagnostics output for channel discovery tests

sveseli committed:
initialize udp sender port

sveseli committed:
add more diagnostics

sveseli committed:
prevent direct connection attempt to unknown port

sveseli committed:
restore logging level

sveseli committed:
resolve codacy complaints

sveseli committed:
reduce test logging level

sveseli committed:
fix setsockopt call for mingw

mdavidsaver commented:
Looking again...

> * Server respects EPICS_PVA_UDP_SENDER_PORT variable; default is 0 (random port), as before

How is `EPICS_PVA_UDP_SENDER_PORT` meant to be used?

Is this intended to address #159?


> * Specifying EPICS_PVA_ADDR_LIST will result with direct tcp connection attempts using the specified addresses, in addition to the regular udp search

EPICS_PVA_ADDR_LIST is defined as a list of UDP endpoints.  How can appropriate TCP ports be deduced from this list without a search phase?

> the PR contains a simple name server utility that is capable of discovering PVA servers on the network and polling them for the list of available channels at a regular intervals.

How would this polling scale wrt. bandwidth usage and maximum reply size?

When I think about something like this, I imagine using a subscription to push incremental changes to a server's PV list.  (similar to what recsync does)

mdavidsaver reviewed (changes requested):
I have still more to say, but am out of time for today.

sveseli commented:
> Looking again...
> 
> > * Server respects EPICS_PVA_UDP_SENDER_PORT variable; default is 0 (random port), as before
> 
> How is `EPICS_PVA_UDP_SENDER_PORT` meant to be used?
> 
> Is this intended to address #159?

This is one of possible solutions for #159, although at this point one can also use tcp searches. By default, udp sender port is arbitrary (0) as before. If you set this variable, this will configure sender port to a specific value.

> > * Specifying EPICS_PVA_ADDR_LIST will result with direct tcp connection attempts using the specified addresses, in addition to the regular udp search
> 
> EPICS_PVA_ADDR_LIST is defined as a list of UDP endpoints. How can appropriate TCP ports be deduced from this list without a search phase?
> 

If udp searches work on the network, there is no difference in behavior as direct connection to the server will not happen. If udp searches do not work, we try to connect directly to servers on the list. If no port is specified as part of the address, we take the default PVA server port.

> > the PR contains a simple name server utility that is capable of discovering PVA servers on the network and polling them for the list of available channels at a regular intervals.
> 
> How would this polling scale wrt. bandwidth usage and maximum reply size?
> 
> When I think about something like this, I imagine using a subscription to push incremental changes to a server's PV list. (similar to what recsync does)

I suppose the intention was that this works in a way that is as unobtrusive as possible, but I  have not done any large scaling tests yet. For APS DAQ system, we have probably on the order of 40-50 PVA servers, which would translate to the same number of "pvlist" calls per polling period, and should result in a relatively low bandwidth usage.  Hopefully, @anjohnson can deploy this soon and see how things work.

sveseli commented on src/remote/pv/channelSearchManager.h:
```diff
@@ -7,18 +7,7 @@
-#   define epicsExportSharedSymbols
-#       undef channelSearchManagerEpicsExportSharedSymbols
-#endif
-
```
So this caused me a lot of headaches, and redoing some of those exports was done following discussions with @anjohnson, whose prescription was a bit simpler than what was in the code. Unfortunately, using *API macro scheme is not easily done for pvAccess, so I opted to go with a minimal number of changes.

If you know of any platform where this does not compile, I can look into it and revise the code.

sveseli replied:
I restored the original epicsExport* statements.

mdavidsaver replied:
> If you know of any platform where this does not compile, I can look into it and revise the code.

The epics-base repository has a much larger windows CI matrix.  I would suggest that you fork the epics-base repository under your account, update modules/pvAccess with this PR, setup appveyor for this fork, and do a test build.

sveseli commented on src/remote/Makefile:
```diff
@@ -3,6 +3,7 @@
 SRC_DIRS += $(PVACCESS_SRC)/remote
 
 INC += pv/security.h
+INC += pv/remote.h
```
pvans utility uses some of the same client code as pvlist, and this was the most expedient solution (the least amount of changes) that I could come up with. If this is really a strong objection, I can try to reshuffle some code around.

sveseli committed:
add missing 2 bytes

sveseli commented on src/remote/channelSearchManager.cpp:
```diff
@@ -196,6 +208,46 @@ void ChannelSearchManager::newServerDetected()
 
+void ChannelSearchManager::send(epics::pvData::ByteBuffer* buffer, TransportSendControl* control)
+{
+    control->startMessage(CMD_SEARCH, 4+1+3+16+2+1+4);
```
You are correct, I will commit the change.

sveseli commented on src/remote/channelSearchManager.cpp:
```diff
@@ -188,6 +191,15 @@ void ChannelSearchManager::searchResponse(const ServerGUID & guid, pvAccessID ci
+    releaseNameServerTransport();
+}
+
+void ChannelSearchManager::releaseNameServerTransport()
```
This is true, on the server side this is not used. Adding this to the "Context" interface follows the pattern used before with similar methods needed on the client side only.

sveseli commented on src/utils/Makefile:
```diff
@@ -12,6 +12,8 @@ INC += pv/wildcard.h
 INC += pv/fairQueue.h
 INC += pv/requester.h
 INC += pv/destroyable.h
+INC += pv/stringUtility.h
```
Okay. Do we have a similar common utility class that I could use instead, perhaps as a part of other modules?

sveseli commented on src/remote/pv/codec.h:
```diff
@@ -303,7 +303,7 @@ class BlockingTCPTransportCodec:
 
     virtual std::string getType() const OVERRIDE FINAL {
-        return std::string("tcp");
+        return PVA_TCP_PROTOCOL;
```
Okay. I find the code easier to read with constants, but I will revert the change and leave "tcp" in place everywhere.

sveseli committed:
revert protocol type back to hardcoded string

sveseli commented on src/remote/codec.cpp:
```diff
@@ -1608,7 +1608,7 @@ void BlockingServerTCPTransportCodec::destroyAllChannels() {
             logLevelDebug,
-            "Transport to %s still has %zu channel(s) active and closing...",
+            "Transport to %s still has %lu channel(s) active and closing...",
             _socketName.c_str(), _channels.size());
```
Sure, I will make the needed format changes.

sveseli commented on src/server/Makefile:
```diff
@@ -4,6 +4,12 @@ SRC_DIRS += $(PVACCESS_SRC)/server
+INC += pv/nameServer.h
+INC += pv/responseHandlers.h
+INC += pv/serverChannelImpl.h
+INC += pv/serverContextImpl.h
```
Hmm, I think nameServer.h was certainly needed to compile pvans utility, but I will take a look at others.

sveseli replied:
This was addressed, no additional headers have been installed as public.

sveseli commented on src/remote/pv/remote.h:
```diff
@@ -185,6 +184,11 @@ class epicsShareClass Transport : public epics::pvData::DeserializableControl {
+    /**
+     * Is transport used
+     */
+    virtual bool isUsed() = 0;
```
In clientContextImpl.

sveseli committed:
use static cast for %lu format

anjohnson commented:
Our CA name-servers currently connect to over 800 IOCs, and we're coming up on 2 million record names. Repeatedly polling probably doesn't scale to that many names or servers and would be a waste of bandwidth and power, although that won't be a problem for us today or in the near future, we aren't planning on building many IOCs with QSRV right now.

My preference would be for the name-server to be able to subscribe to each server's pvlist and get updates when PVs are added or removed. That obviously isn't necessary for the current IOC with QSRV, but other PVA servers can already be dynamic, and we might allow IOCs to add, remove or rename records at some point.

The other change to avoid the need for polling is to have an API for clients to discover when a new beacon (or a beacon anomaly in CA parlance) is seen, indicating a new server coming online. The name-server could then introspect new servers as soon as it sees them, and there's no polling delay before it knows about them. I believe PVA's beacon behavior is slightly different than CA though, is it as reliable as CA in the event that a network segment gets disconnected and reconnected later without the servers beyond it rebooting?

sveseli committed:
move stuff around to avoid exposing additional headers

sveseli committed:
few more changes for mingw

sveseli committed:
fix codacy complaint about explicit constructors

sveseli committed:
restore original export defs; remove unused constants

sveseli committed:
fix format warning

sveseli committed:
add test disgnostics output

sveseli committed:
new test added

mdavidsaver commented:
I think this PR is getting too large.  With several distinct changes.

Adding support for `EPICS_PVA_NAME_SERVERS` is a reasonable addition, and I think should be uncontentious.  __Can this change be extracted as a separate PR?__

As for the "Direct Connection" change(s).  Do I understand correctly that the goal here is to bypass the search phase entirely.  That is, send CREATE_CHANNEL without first receiving a positive SEARCH_REPLY?

If so, I can understand fixing bugs in the client so that eg. direct channels created through the existing API will reconnect automatically.

Adding support for URI-like direct connect syntax to pvget and friends might also be acceptable.  (I am not enthusiastic, but I suspect at least one person would be)

However, I think that beginning to treat `EPICS_PVA_ADDR_LIST` as a list of TCP addresses would be unacceptable.  The `*_ADDR_LIST` variables for both PVA and CA have always been lists of UDP endpoints, including multicast and broadcast addresses.  Changing this definition now seems to me likely to cause problems with other code and/or user expectations.

wrt. `EPICS_PVA_UDP_SENDER_PORT`.  I am not sure this would actually help with #159, and would anyway be a burden on sysadmins.  I think a proper solution to #159 is for a server to always a send search reply back through the same socket which received that search request.  This should be a (tedious) exercise in interface re-plumbing.  One which I think will anyway be necessary to handle `EPICS_PVA_NAME_SERVERS` (search over TCP must reply over the same connection).

wrt. `pvans`, I think it is too soon to include something like this in pvAccessCPP.  It sounds to me like this tools is at the stage of small scale prototype.

anjohnson commented:
@mdavidsaver We will do this:

1. Comment out the direct connection code, and using EPICS_PVA_ADDR_LIST as a list of TCP addresses. The direct connection code was already present, and reconnections did work using it, but we don't actually need it.
2. Separate out the name-server program and internal classes to implement the new API it uses. This would become a separate PR.
3. What would it take to either allow or drop the `EPICS_PVA_UDP_SENDER_PORT` parameter — issues #159 and #128 may have said "here be dragons" on duplicating the port number, how to discover it what's a real problem? Would you be happy to keep it in if we just changed the default value from zero (use a random reply port) to the value of the `EPICS_PVA_BROADCAST_PORT` parameter?

Is that likely to be acceptable, or are there more changes you'd like to see to merge this functionality?

Thanks, Andrew & Sinisa

mdavidsaver commented:
> 2\. name-server program ... the new API it uses

A proposed design of any new/extended APIs seems like a good starting point.  eg. beginning with the question of where the current ChannelProvider interface is insufficient?

Also, I would strongly encourage that Sinisa and Co. study the design of [recsync](https://github.com/ChannelFinder/recsync), which is able to avoid eg. repeatedly polling the record list by maintaining a persistent connection with each server.

fyi. redesigning recsync around PVA in addition to (and eventually instead of) my custom protocol would be a useful improvement.  Since it is in active use, the pluses and minuses of that data model are known.

mdavidsaver commented:
> 3\. What would it take to either allow or drop the `EPICS_PVA_UDP_SENDER_PORT` parameter ...

Maybe we could have an example with Linux (iptables or nft)?  How can this new parameter be used to configure a firewall on a host with a variable number of PVA clients and server?  eg. must each client process be manually assigned a unique sender port?


> ... may have said "here be dragons" on duplicating the port number ...

These "dragons" lurk in the dark sprawling internal APIs of pvAccessCPP, and present an entirely artificial barrier between the code wishing to send a UDP search reply, and the socket on which the corresponding search request was received.

imo. there is no need for a separate `sendTransport` socket.  The best course of action (short of using PVXS) would be to eliminate that object, redesigning the internal API so that UDP replies are always send through the socket which received the request.

A painful task, but I think preferable to burdening every admin with additional configuration.

sveseli committed:
remove pvans utility

sveseli committed:
remove nameserver support classes

sveseli committed:
remove string utility tests

sveseli committed:
remove ability to set udp sender port

sveseli committed:
disable code that initiates direct connection to server based on EPICS_PVA_ADDR_LIST

sveseli committed:
remove search acceptor; name server searches have to go through the regular server port

sveseli committed:
remove unused method

sveseli committed:
add default implementation of isUsed()

sveseli committed:
add default implementation name server search methods

sveseli commented:
> Please add some test converge.

Tests have been added.

sveseli commented:
> @mdavidsaver We will do this:
> 
>     1. Comment out the direct connection code, and using EPICS_PVA_ADDR_LIST as a list of TCP addresses. The direct connection code was already present, and reconnections did work using it, but we don't actually need it.
> 
>     2. Separate out the name-server program and internal classes to implement the new API it uses. This would become a separate PR.
> 
>     3. What would it take to either allow or drop the `EPICS_PVA_UDP_SENDER_PORT` parameter — issues [PVA Server's use of random 'sendTransport' UDP port makes it impractical for gateway #159](https://github.com/epics-base/pvAccessCPP/issues/159) and [Client and server unintentionally sharing a UDP source port #128](https://github.com/epics-base/pvAccessCPP/issues/128) may have said "here be dragons" on duplicating the port number, how to discover it what's a real problem? Would you be happy to keep it in if we just changed the default value from zero (use a random reply port) to the value of the `EPICS_PVA_BROADCAST_PORT` parameter?
> 
> 
> Is that likely to be acceptable, or are there more changes you'd like to see to merge this functionality?
> 
> Thanks, Andrew & Sinisa

* Commented out the code that forced direct tcp connection to the server
* Removed ability to set udp sender port
* Removed pvans utility and name server related classes

The only thing that is left as part of the PR is the support for tcp channel searches via EPICS_PVA_NAME_SERVERS variable. If this is not set, the code should behave as before.

shroffk commented:
> Extend the channel finder's IOC tool (reccaster?) to provide the TCP port. I think that's been completed.
> Channel finder will thus not only know the PV names but also the TCP IP and port.
> Next build a name server that starts like that demo but instead of taking info for 1 or 2 PVs from the command line it queries the channel finder.

Hello,

reccaster has been updated to include the TCP port which is not exported as a environment variable. We need to standardize on the ChannelFinder property structure ( this is technically simple just needs some co-ordination)

A prototype name serve which replies to queries using the above info from ChannelFinder is available here
https://github.com/ChannelFinder/cfNameserver
I have tested this locally and it works fine... the next steps will be updating the service to match the agreed upon ChannelFinder name property and stress testing for performance.

shroffk commented:
Sorry about the accidental close

anjohnson commented:
Core Group: MAD will review this and provide feedback before the next meeting.

sveseli committed:
remove pvaConstants.cpp, restore pvaVersion.cpp, remove PVA_UDP_SENDER_PORT constant

sveseli committed:
remove udp sender port variables

sveseli committed:
remove unused utility method for converting string to inet address

sveseli committed:
remove unused methods, restore original initialize() method for server context impl class

sveseli committed:
remove unused server search response handler class

sveseli committed:
reduce logging for request start

sveseli committed:
force name server transport release if channel is not found after max nubmer of attempts

mdavidsaver commented:
With @sveseli s recent pruning, I think I see what is going on wrt. name server handling.  A couple of questions to check my understanding.

Why are connections to name servers treated differently to other connections?

Is each client context limited to connecting to one name server at a time?

wrt. `releaseNameServerTransport()`.  This looks like manual ref. counting.  Can it be avoided?

wrt. the design around `getNameServerSearchTransport()`.  Can you avoid looping for blocking `connect()` from the search timer callback?

eg. doing so stalls searching while name server(s) are offline.

```
$ time ./bin/linux-x86_64/pvget foo
Timeout
foo 
real    0m5.029s
$ time EPICS_PVA_NAME_SERVERS=10.127.127.2 ./bin/linux-x86_64/pvget foo
Timeout
foo 
real    0m9.227s
$ time EPICS_PVA_NAME_SERVERS="10.127.127.2 10.127.127.3" ./bin/linux-x86_64/pvget foo
Timeout
foo 
real    0m12.285s
```

Compare with:

```
$ time EPICS_PVA_NAME_SERVERS="10.127.127.2 10.127.127.3" ./bin/linux-x86_64/pvxget foo
2024-05-19T14:17:47.878231057 ERR pvxs.tcp.io connection to Server 10.127.127.3:5075 closed with socket error 113 : No route to host
2024-05-19T14:17:47.878889508 ERR pvxs.tcp.io connection to Server 10.127.127.2:5075 closed with socket error 113 : No route to host
Timeout with 1 outstanding

real    0m5.031s
```

sveseli committed:
remove blocking call for getting name server connections, and instead use timers; if multiple name servers are configured, connections to all of them are obtained at the same time

sveseli committed:
do not release all name server connections if channel search was unsuccessful

sveseli commented:
The blocking call for name server connections has been removed. Name server connections now use separate set of timers, so that they can be established at the same time. In the examples below I used non-existent hosts:

<pre>
$ time pvget foo
Timeout
foo 
real	0m5.021s
user	0m0.011s
sys	0m0.015s

$ time EPICS_PVA_NAME_SERVERS="192.168.0.112:11111" pvget foo
Timeout
foo 
real	0m5.128s
user	0m0.010s
sys	0m0.023s

$ time EPICS_PVA_NAME_SERVERS="192.168.0.112:11111 192.168.0.113:22222" pvget foo
Timeout
foo 
real	0m5.113s
user	0m0.000s
sys	0m0.021s

</pre>

Client can now establish connections to multiple name servers at the same time, and they are released as soon as they are no longer needed or aren't useful (hence methods that release those connections). Name server connections reuse existing classes (e.g. TransportRegistry, BlockingTCPConnector, etc.) as much as possible.

sveseli commented:
@mdavidsaver Is there anything else you would like to see done or modified before this PR can be merged?

sveseli commented:
@mdavidsaver @anjohnson Is this PR okay to be merged, or would you like to see some other changes? I would like to make sure that this goes into the next release, if that would be possible.

mdavidsaver commented:
> ... Is this PR okay to be merged ...

No.  There are several aspects of this proposed design which I do no like.  eg. blocking `connect()` not on a dedicated thread.  Pointing these out piecemeal does not seem to be moving us towards a resolution, but I have not found the time to write out an alternate design.

sveseli commented:
> > ... Is this PR okay to be merged ...
> 
> No. There are several aspects of this proposed design which I do no like. eg. blocking `connect()` not on a dedicated thread. Pointing these out piecemeal does not seem to be moving us towards a resolution, but I have not found the time to write out an alternate design.

The most recent commits (two months ago) addressed the issue of blocking connect you brought up. The connections are done via a separate set of timers. If there are other issues that need to be resolved, I would be happy to do this.

mdavidsaver commented:
> The most recent commits (two months ago) addressed the issue of blocking connect you brought up. ...

Then what is going on with this `wait(0.1)`?

```c++
           // Wait to get transport if we do not have it
            m_nsTransportEvent.wait(MAX_NS_TRANSPORT_WAIT_TIME);
```

As for the rest.  I have a limited amount of time which I can spend on EPICS community work, and many projects which ask for my attention.  I would have an easier time reviewing your proposed changes if you would say more about what you are trying to achieve, and how you intend to achieve it.  Correctly inferring intent from code is both difficult and error prone.

> ... released as soon as they are no longer needed ...

eg.  Do you intend that an idle client will not maintain name server connections?  Why not maintain NS connections continually?  Are you are making a trade off of resource use vs. (re)connect latency?

eg.  Why are connections to name servers treated differently to other PVA server?  Do you see this as a "need", a convenience, ...?

sveseli commented:
> > The most recent commits (two months ago) addressed the issue of blocking connect you brought up. ...
> 
> Then what is going on with this `wait(0.1)`?
> 
> ```c++
>            // Wait to get transport if we do not have it
>             m_nsTransportEvent.wait(MAX_NS_TRANSPORT_WAIT_TIME);
> ```
> 

If you look closely, the earlier code had an actual connect() in this thread. Right now a name server connect timer is scheduled that does the actual connect, and notifies corresponding event if connection was established. I used a short wait on this event just to see if connection can be established quickly, which improves search performance greatly when the name server responds, and does not hurt very much if name server does not respond, as we do not have channel connection anyways. I could certainly make sure that this wait happens only when the name server connection timer starts, rather than on every name server search attempt, but I suspect this will have a very little effect in terms of channel timeouts, etc. (see my examples of what happens when trying to connect to channels using non-existent machines).  


> As for the rest. I have a limited amount of time which I can spend on EPICS community work, and many projects which ask for my attention. I would have an easier time reviewing your proposed changes if you would say more about what you are trying to achieve, and how you intend to achieve it. Correctly inferring intent from code is both difficult and error prone.
> 
> > ... released as soon as they are no longer needed ...
> 
> eg. Do you intend that an idle client will not maintain name server connections? Why not maintain NS connections continually? Are you are making a trade off of resource use vs. (re)connect latency?
> 

This code is intended to work for a general case, where the name server connection is different from PVA Server connections. Rather than every client maintaining name server connections (and thus keeping those TCP sockets alive), I chose to close those connections as soon as I did not need them in order to preserve resources. There is obviously a trade off here in terms of resources and re-connection latency. One can perhaps argue one way or another as to what is more important, but given that name server can have hundreds or even thousands of client connections at any given time, I decided to prioritize resources, rather than to worry about reconnecting a bit faster every once in a while.

> eg. Why are connections to name servers treated differently to other PVA server? Do you see this as a "need", a convenience, ...?

This stems partly from the fact that in general connections to name servers are different from connections to PVA servers in the sense we use them only in the discovery phase and also (for the reasons mentioned above) I chose to close them as soon as possible. Another reason is trying to minimize amount of the new code and impact of the required changes to the existing code, and making sure that no matter what happens with the name server connections, those do not affect client behavior and existing PVA server connections.

I do appreciate the feedback/suggestions and I am willing to make any changes that are feasible and do not require major rewrite of the existing code.

sveseli committed:
wait for name server transport only if the search has been just scheduled

anjohnson commented:
Meeting:
MD: All servers must support "_server" for RPC to get name list.
SV: This PR now just adds support for PVA searches over TCP, it doesn't implement a name-server. That will come in a separate PR (that was removed from this PR earlier).
Agreed: MD will see implementation of a future name server before SV distributes it.
MD: `QOS_REPLY_REQUIRED` doesn't need to be set, a searching application cancelling the search is the only thing that can cause the number of searches to go to 0. This should reduce traffic. Careful to avoid stalling the search process if the name-server dies.
SV: Client only sends 3 searches to a connected name-server before giving up and trying the next name-server in its list.
ANJ: "The nameserver" — we want to be able to query multiple name-servers at the same time, to reduce search latency.
SV: Want to save resources by not keeping sockets open when we don't have any names we haven't found.
MD: Having multiple open sockets is likely to need less resources than multiple threads.
ANJ: Okay with closing NS sockets when all searches done, but wait for some time period after the "last name found" for the client to ask for more channels.
MD: Define test cases to prove that this feature is doing what is intended. Measure that UDP searches and TCP searches don't take significantly different periods of time to complete. How long does it take to search for 10,000 PVs over both transports? Use the client API to measure those, so not a micro-benchmark.

AIs on SV:
- Client searches should be sent to multiple NSs in parallel.
- Wait for a short time after searching before closing TCP socket to the NS, in case client is inefficient and does (search; wait; get; wait) multiple times.
- Measure how long connections, and later reconnections, of 10,000 channels take through both transports. Not a unit-test, report the results in a comment here (may prompt further discussion), can use pvaPy. Specifically connect to 5 PVs, wait for all connect and for the NS to disconnect, then connect to 10,0000 PVs and measure how long it takes. Check separately against both 1 and 2 name-servers, with some channels in both. Test for (everything UDP NS; everything TCP NS; split between 2 UDP NSs; split between 2 TCP NS; split between 1 UDP and 1 TCP NS).

sveseli committed:
Merge branch 'epics-base:master' into master
